### PR TITLE
docs: evm-machine design doc

### PR DESCRIPTION
**Pull request by enitrat**
## Pull Request type

In this design document, we dive into a refactor of the Kakarot's EVM-Machine implementation, compatible with Cairo's limitations raised in https://github.com/starkware-libs/cairo/issues/4032. 
The accepted proposed design will need to: 
- Be easily refactorable
- Have a low overhead cost compared to the one originally planned
- Not introduce development complexity in other parts of the code

I believe that the proposed one performs well in all the points mentioned above. Please discuss it extensively!

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [x] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Review by Eikix — COMMENTED** [review]
What are the limitations of this approach?

**Comment by enitrat** [comment]
> What are the limitations of this approach?

It's a bit more expensive as we need to track each stack/memory's context-related segment length in a dict instead of as a simple variable.
Also, this imposes a limit on the maximum amount of memory space a context can allocate.

**Comment by Eikix** [comment]
> > What are the limitations of this approach?
> 
> 
> 
> It's a bit more expensive as we need to track each stack/memory's context-related segment length in a dict instead of as a simple variable.
> 
> Also, this imposes a limit on the maximum amount of memory space a context can allocate.

Let's discuss it in our weekly
What is the limit on memory space per context?

**Comment by enitrat** [comment]
> the gas costs required to have a memory containing
> 125000 words is above the 30M gas limit. We will use this heuristic to bound the
> maximum size of the memory to 125000 256-bits words. Therefore, we will bound
> the maximum size of the memory to 125000 256-bits words.

**Comment by enitrat** [comment]
@Eikix If you're okay with it I propose to merge the current design proposal, and to apply any modifications from SW's feedback in a new PR

File: docs/general/model/execution_context.md [hunk 1]
```diff
@@ -26,9 +26,11 @@ classDiagram
     stopped: bool,
     +gas_limit: u64
     +gas_price:u64
-    +memory Memory
-    +stack Stack
-    +u32 program_counter
+    +memory: Memory
```
Comment (Eikix) [inline docs/general/model/execution_context.md:29]: it seems this data structure does not match the data structure below:
- machine
- executioncontext 
etc.

Comment (enitrat) [reply docs/general/model/execution_context.md:29]: Ah, yes, indeed, thanks for pointing this out, I need to rework this section as well

**Comment by enitrat** [comment]
@Eikix Reworked this:
- Changes brought to the ExecutionContext docs
- The Storage section has been entirely reworked

File: docs/general/machine.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+## The Kakarot Machine design
+
+To overcome the problem stated above, we have come up with the following design:
+
+- There is only one instance of the Memory and the Stack, which is shared
+  between the different execution contexts.
+- Each execution context has its own identifier `id`, which uniquely identifies
+  it.
+- Each execution context has a `parent_context` field, which is the numerical
```
Comment (Eikix) [inline docs/general/machine.md:29]: is this true? the classDiagram says Nullable<ExecutionContext>

File: docs/general/machine.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+  identifier of its optional parent execution context.
+- Each execution context has a `child_context` field, which is the numerical
```
Comment (Eikix) [inline docs/general/machine.md:31]: same q

File: docs/general/machine.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+Considering Cairo's limitations raised previously, we will use a single data
+structure to track local and global storage changes. We will use a `Journal`
+data structure, that will track two things: the changes performed in the current
+execution context, and the changes performed in the transaction as a whole. The
+Journal will have the following fields:
+
+```rust
+  struct Journal {
+      local_changes: Felt252Dict<felt252>,
```
Comment (Eikix) [inline docs/general/machine.md:195]: what if the local changes are u256?

Comment (Eikix) [reply docs/general/machine.md:195]: https://github.com/Eikix/kakarot/blob/aec81a483c3862bd71e0d55850a64f338434b4bc/src/kakarot/accounts/contract/library.cairo#L25

File: docs/general/machine.md [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+    class CallContext{
+        caller: EthAddress,
+        value: u256,
+        bytecode: Span~u8~,
+        calldata: Span~u8~,
+    }
+
+    class Journal{
+        local_changes: Felt252Dict~felt252~,
```
Comment (Eikix) [inline docs/general/machine.md:92]: should journal expand its logic to `created_contracts` and `destroyed_contracts`?
as well as contract account nonce (if the contract account runs many create tx, nonce should be optimistically increased, until tx is finalized, and we finally write the final nonce to storage)

File: docs/general/machine.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        origin: EthAddress,
+        call_ctx: CallContext,
+        gas_price: u32,
+        gas_limit: u32,
+        pc: u32,
+        read_only: bool,
+        destroyed_contracts: Array~EthAddress~,
+        events: Array~Event~,
+        create_addresses: Array~EthAddress~,
+        reverted: bool,
```
Comment (Eikix) [inline docs/general/machine.md:77]: should reverted / stopped be a status enum?

File: docs/general/machine.md (hunk 4 above)
Comment (enitrat) [reply docs/general/machine.md:195]: Yes they're actually always u256 indeed

File: docs/general/machine.md (hunk 6 above)
Comment (enitrat) [reply docs/general/machine.md:77]: like
```rust
enum Status{
Active,
Stopped,
Reverted
}
```
?

Comment (Eikix) [reply docs/general/machine.md:77]: Yeah

**Comment by enitrat** [comment]
@Eikix Changes applied

**Review by Eikix — APPROVED** [review]
lgtm 🔥

**Review by Eikix — CHANGES_REQUESTED** [review]
mb just noticed this PR does not have mentions of "dirtyStorage"? 
is dirtyStorage the equivalent of `local_changes`?
should we modify the other pr #360

**Comment by enitrat** [comment]
>mb just noticed this PR does not have mentions of "dirtyStorage"?
>is dirtyStorage the equivalent of local_changes?
>should we modify the other pr https://github.com/kkrt-labs/kakarot-ssj/pull/360

dirtyStorage is a name that is unique to Geth. For us, it's the combination of `local_changes` and `global_changes`.

**Review by Eikix — APPROVED** [review]

**Pull request merged**